### PR TITLE
Topic/first real example

### PR DESCRIPTION
**Pull request by vjovanov**
This pull request rounds up on running one branch of the first example with reflection.

### CODE REVIEW

**Comment by xeno-by** [comment]
Hello, @vjovanov! Thank you for your interest in contributing to the scala.meta project. Please sign the [CLA](http://typesafe.com/contribute/cla/scala), so that we can proceed with reviewing your pull request.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 1]
```diff
@@ -0,0 +1,288 @@
+package scala.meta.internal
+package interpreter
+
+import scala.util.matching._
+import scala.meta._
+import scala.meta.semantic._
+import scala.meta.internal.{ ast => i }
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Later on we might want to unify our notations for `scala.meta.internal.ast`. In scalameta/scalameta, it's `impl.`, in scalameta/scalahost, it's `m.`, in Intellij's host it's `m.`.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+import scala.collection.immutable.ListMap
+import scala.meta.internal.interpreter.environment._
+
+object Interpreter {
+
+  def evalFunc(metaprogram: i.Tree, argss: Seq[Any]*)(implicit c: Context): Any =
+    evalFunc(metaprogram, argss, Env(List(ListMap[i.Term.Name, Object]()), ListMap[i.Term.Name, Object]()))._1
+
+  def eval(term: Term)(implicit c: Context): Any = {
+    val (Object(v, tpe, fields), env) =
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Can the result end up being not `Object` here?

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+      case nme @ i.Term.Name(_) =>
+        (env.lookup(nme), env)
+
+      // TODO dispatch to pattern matching in an elegant way
+      case i.Defn.Val(mods, List(i.Pat.Var.Term(name)), tpe, body) =>
+        val (res, newEnv) = eval(body, env)
+        // TODO depends of the context (method body or template)
+        (Object((), t"Unit"), newEnv.push(name, res))
+
+      // ASK Eugene if we can replace them with one!
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Not at the moment.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      case i.Lit.Unit() =>
+        (Object((), t"Unit"), env)
+
+      case i.Term.Tuple(terms) =>
+        val (res, env1) = evalSeq(terms, env)
+        (Object((res.head.ref, res.tail.head.ref), t"(AnyRef, AnyRef)"), env1) // TODO other cases
+
+      case i.Term.Select(lhs, op) =>
+        val (lhsV, env1) = eval(lhs, env)
+        val method = lhs.tpe.members.filter(_.name == op).head
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: I think you should be able to do `members(op)` here instead of `members.filter(_.name == op).head`.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        val (lhsV, newEnv) = eval(lhs, env)
+        methodCall(method, lhsV, Seq(), newEnv)
+      case i.Term.ApplyInfix(lhs, op, _, args) =>
+        val method = lhs.tpe.members.filter(_.name == op).head
+        val (lhsV, newEnv) = eval(lhs, env)
+        val (rhsVs, callEnv) = evalSeq(args, env)
+        methodCall(method, lhsV, rhsVs, callEnv)
+      case i.Term.Apply(i.Term.Select(lhs, rhs), args) =>
+        val (lhsV, newEnv) = eval(lhs, env)
+        val method = lhs.tpe.members.filter(_.name.toString == rhs.name.toString).head
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Why did you need toString here?

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        val (rhsVs, callEnv) = evalSeq(args, env)
+        methodCall(method, lhsV, rhsVs, callEnv)
+      case i.Term.Apply(lhs, args) =>
+        val method = lhs.tpe.members.filter(_.name.toString == "apply").head
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: I think that `defs("apply")` would work here.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+      case _ => Utils.unsupported(term, "term")
+    }
+  }
+
+  def evalPattern(lhs: Object, pattern: i.Pat, env: Env)(implicit c: Context): (Object, Env) = pattern match {
+    case i.Pat.Typed(x, tp) =>
+      (Object(tp.tpe <:< lhs.tpe, t"Boolean"), extendEnv(x, lhs, env))
+    case i.Pat.Wildcard() =>
+      (Object(true, t"Boolean"), env)
+    case _ => Utils.unsupported(pattern, "pattern")
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: In scalameta-foundation, there's `org.scalameta.unreachable` or `org.scalameta.require`.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+    lhsJTp: String, nme: String, argsRetJTp: String, env: Env)(implicit c: Context) = {
+
+    val paramsRegex = """\((.*)\).*""".r
+    val paramsRegex(jvmParams) = argsRetJTp
+    val types = jvmParams.replaceAll("(L.*?;|I|B)", "$1,").split(",").toList.filterNot(_.trim.isEmpty).map(Utils.jvmTypeToClass)
+    println("Expected class is: " + Utils.jvmTypeToClass(lhsJTp))
+    val jMethod = Utils.jvmTypeToClass(lhsJTp).getMethod(nme, types: _*)
+
+    val repeated = params.exists(_.exists {
+      case i.Term.Param(_, _, Some(i.Type.Arg.Repeated(_)), _) => true
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: You can use `param.isVarargParam` here.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+          Seq(context.asInstanceOf[AnyRef])
+        case ("Lscala/meta/semantic/Api/XtensionSemanticMemberLike;", "children", "(Lscala/meta/semantic/Context;)Lscala/collection/immutable/Seq;") =>
+          val context = env.stack.flatten.find(_._1.toString == "c").head._2.ref
+          Seq(context.asInstanceOf[AnyRef])
+        case ("Lscala/meta/semantic/Api/XtensionSemanticMemberLike;", "isFinal", "(Lscala/meta/semantic/Context;)Z") =>
+          val context = env.stack.flatten.find(_._1.toString == "c").head._2.ref
+          Seq(context.asInstanceOf[AnyRef])
+        case ("Lscala/meta/semantic/Api/XtensionSemanticMemberLike;", "isCase", "(Lscala/meta/semantic/Context;)Z") =>
+          val context = env.stack.flatten.find(_._1.toString == "c").head._2.ref
+          Seq(context.asInstanceOf[AnyRef])
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Sorry for not having implemented `desugar` yet.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+  }
+}
+
+object Utils {
+  def jvmTypeToClass(s: String): Class[_] = s match {
+    case "I" => classOf[Int]
+    case "Lscala/meta/semantic/Api/XtensionSemanticTypeRefDefn;" => // TODO URGH!      
+      Class.forName("scala.meta.semantic.Api$XtensionSemanticTypeRefDefn")
+    case "Lscala/meta/semantic/Api/XtensionSemanticMemberLike;" => // TODO URGH!
+      Class.forName("scala.meta.semantic.Api$XtensionSemanticMemberLike")
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Looks like a bug in `jvmsig`. I'll look into this right now.

File: project/build.scala [hunk 11]
```diff
@@ -124,11 +125,11 @@ object build extends Build {
   ) settings (
     test in Test := (test in tests in Test).value,
     packagedArtifacts := Map.empty
-  ) aggregate (interpreter, tests)
+  ) aggregate (interpreter_, tests)
 
-  lazy val interpreter = Project(
-    id = "interpreter",
-    base = file("interpreter")
+  lazy val interpreter_ = Project(
```
Comment (xeno-by) [inline project/build.scala]: Why the underscore?

File: project/build.scala [hunk 12]
```diff
@@ -154,6 +155,26 @@ object build extends Build {
     libraryDependencies += "org.scalatest" %% "scalatest" % "2.1.3" % "test",
     libraryDependencies += "org.scalacheck" %% "scalacheck" % "1.11.3" % "test",
     libraryDependencies += "org.scalameta" % "scalameta_2.11" % "0.1.0-SNAPSHOT",
+    libraryDependencies += "org.scalameta" % "scalahost_2.11.6" % "0.1.0-SNAPSHOT" % "test",
+    testOptions in Test += Tests.Argument("-oDF"),
```
Comment (xeno-by) [inline project/build.scala]: What does this stand for?

File: project/plugins.sbt [hunk 13]
```diff
@@ -1 +1,3 @@
 addSbtPlugin("com.typesafe.sbt" % "sbt-pgp" % "0.8.1")
+
+addSbtPlugin("com.typesafe.sbt" % "sbt-scalariform" % "1.3.0")
```
Comment (xeno-by) [inline project/plugins.sbt]: Just wondering. How's this used?

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 2 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: No, it is always an object. There is no distinction between primitive and non-primitive values.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 3 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Oki, I will just write them all then for completeness.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 8 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: That did not work at the moment I used it. I will try again.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 9 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: :) This works as well.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 10 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Yep!

File: project/build.scala (hunk 12 above)
Comment (vjovanov) [inline project/build.scala]: Prints the full stack trace (F) and the duration of each test (D).

File: project/plugins.sbt (hunk 13 above)
Comment (vjovanov) [inline project/plugins.sbt]: All the code in this project is formatted with scalariform.

**Comment by vjovanov** [comment]
Do not merge this yet. I want to make both examples work and then we can merge.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 1 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Agree. I am good with all conventions as long as it is short.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 4 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: This was a temporary hack since for `+` when I do members I get the following exception:
```
- should deal with conditionals *** FAILED *** (45 milliseconds)
[info]   scala.meta.SemanticException: multiple term members named "+" found in x.type
[info]   at scala.meta.semantic.Api$XtensionSemanticScopeLike$class.internalSingleNamed(Api.scala:479)
```

How do I get the exact member?

Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: What code leads to that exception?

Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: `if(true) {val x = 1; x + 1} else {val y = 1; y + 1}` 

The else than branch in this case.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 10 above)
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: https://github.com/scalameta/scalahost/pull/75

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 4 above)
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Do you write `members(op)` or `members(op.toString)`?

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 5 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: At the time being the simple equality check did not work for me. Now it works fine.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 6 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Indeed!

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 4 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: `members(op)`. Changing to it fails 4 tests: 
This is the code:
```
case i.Term.ApplyInfix(lhs, op, _, args) =>
    val method = lhs.tpe.members(op)
    val (lhsV, newEnv) = eval(lhs, env)
    val (rhsVs, callEnv) = evalSeq(args, env)
    methodCall(method, lhsV, rhsVs, callEnv)
```

File: project/build.scala (hunk 11 above)
Comment (vjovanov) [inline project/build.scala]: Trying to make it work with Eclipse. I think it is not necessary, however my project folder must be named differently than `interpreter`. I will try to make this project work with Eclipse.

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 7 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: They do not print the semantics and code of trees. Should I modify them with the functionality I need?

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 10 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Thx, the fix works!

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala (hunk 7 above)
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: It's possible to do something `unreachable(debug(pattern, pattern.show[Semantics]))`. But don't worry about this for now.

Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Done! Less code is good.

**Comment by vjovanov** [comment]
Please look at the test cases and see if you would like to add something else. 

Otherwise, the interpreter works.

File: core/src/main/scala/scala/meta/internal/interpreter/Env.scala [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            Object(clazz.getField("MODULE$").get(clazz), nme.tpe)
+        }
+      } else {
+        ??? // TODO then in the scope of current objects
+      }
+
+    }
+  }
+
+  def toClass(nme: TName)(implicit c: Context): Class[_] = {
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Env.scala]: Would it help if I also added `@ffi` on Defn.Class, Defn.Trait, etc?

File: project/build.scala (hunk 11 above)
Comment (xeno-by) [inline project/build.scala]: Btw isn't the changed name also going to affect the name of the produced artifact?

File: core/src/main/scala/scala/meta/internal/interpreter/Engine.scala [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+    }
+  }
+
+  def methodCallByMods(mods: Seq[i.Mod], params: Seq[Seq[i.Term.Param]], lhs: Object, rhs: Seq[Object], env: Env)(implicit c: Context) = {
+    val jvmMethod: String = mods.collect({ case i.Mod.Ffi(s) => s }).head
+    val rgx = """(.*)\((.*), (.*), (.*)""".r
+    jvmMethod.substring(0, jvmMethod.length - 1) match {
+      case rgx("scalaIntrinsic", lhsJTp, nme, argsRetJTp) =>
+        ScalaEmulator.emulate(List(lhsJTp, nme, argsRetJTp), lhs +: rhs, env)
+      case rgx("jvmMethod", "Ljava/lang/Object;", "$plus", "(Ljava/lang/String;)Ljava/lang/String;") => // ARGH!
```
Comment (xeno-by) [inline core/src/main/scala/scala/meta/internal/interpreter/Engine.scala]: Looks like this should be an intrinsic, right?

File: core/src/main/scala/scala/meta/internal/interpreter/Env.scala (hunk 14 above)
Comment (vjovanov) [inline core/src/main/scala/scala/meta/internal/interpreter/Env.scala]: Sure, but this phase is finished now, and we will anyhow go for non-reflective interpretation.

File: project/build.scala (hunk 11 above)
Comment (vjovanov) [inline project/build.scala]: Yes, I do not know how to solve this issue. Wait for this fix and we can merge.

File: project/plugins.sbt (hunk 13 above)
Comment (vjovanov) [inline project/plugins.sbt]: Removing this as we want the interpreter code to be beautifully polished.
